### PR TITLE
Add a new temporal mapping optimizer, SALSA: Simulated Annealing based Loop-Ordering Scheduler for DNN Accelerators

### PR DESCRIPTION
**Pull request by Victor-Jung**

### CODE REVIEW

**Comment by asyms** [comment]
Hi,

Thanks for putting in the effort to merge this. A couple of comments:

- Could you merge in the changes I pushed today?
- Could you change the logger 'info' statements to 'debug' statements? It's annoying for a lot of processes. Maybe just have one info statement that says we're running in salsa mode with X cpu cores?
- Maybe not for this PR, but it would be great if you can add another stage that automatically decides to use loma or salsa based on the layer and settings, like we had in the past.

Thanks!

**Comment by Victor-Jung** [comment]
Hello @asyms,

I did the first 2 points you mentioned and will open a new PR with the TMOptSelectStage later.

Best,
Victor

File: main_onnx_salsa.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#  
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an AS IS BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#  
+
+from zigzag.classes.stages import *
+from zigzag.classes.stages.SalsaStage import *
```
Comment (asyms) [inline main_onnx_salsa.py]: Please import the SalsaStage in zigzag.classes.stages

File: zigzag/classes/stages/SalsaStage.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    def compare_cme_latency(self, cme):
+        """
+        Compare the latency of the current cost model evaluation with the best latency found so far.
+        Then replace the current best cme if the current cme has a lower latency.
+        """
+
+        if self.best_cme is None:
+            self.best_cme = cme
+        elif cme.latency_total0 < self.best_cme.latency_total0:
```
Comment (asyms) [inline zigzag/classes/stages/SalsaStage.py]: Also add that if the latencies are equal but the new cme's energy is better, we pick the new cme as best cme.

File: zigzag/classes/stages/SalsaStage.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+
+    def compare_cme_energy(self, cme):
+        """
+        Compare the energy of the current cost model evaluation with the best energy found so far.
+        Then replace the best cme if the current cme has a lower energy.
+        """
+        if self.best_cme is None:
+            self.best_cme = cme
+        elif cme.energy_total < self.best_cme.energy_total:
```
Comment (asyms) [inline zigzag/classes/stages/SalsaStage.py:154]: Also add that if energies are equal but the new cme's latency is better, we pick the new cme as best cme.

File: zigzag/classes/stages/SalsaStage.py (hunk 2 above)
Comment (asyms) [reply zigzag/classes/stages/SalsaStage.py]: Also, in general we consider latency_total2 (including on- and off-loading), any reason you chose latency_total0 here?

**Review by asyms — CHANGES_REQUESTED** [review]
Added a couple of small comments.

File: zigzag/classes/stages/SalsaStage.py (hunk 2 above)
Comment (Victor-Jung) [reply zigzag/classes/stages/SalsaStage.py]: Didn't realize there was a latency_total2, I picked the first suggestion.

File: main_onnx_salsa.py (hunk 1 above)
Comment (Victor-Jung) [reply main_onnx_salsa.py]: Done

**Review by asyms — APPROVED** [review]
LGTM!

**Pull request merged**